includes/discovery/ports.inc.php

```diff
@@ -108,6 +109,64 @@
+        // We only need to consider results that have an entry in the database
+        if (array_key_exists($ifIndex, $ports_ifindex)) {
+            $snmp_data['ifIndex'] = $ifIndex; // Store ifIndex in port entry
+            $snmp_data['ifAlias'] = StringHelpers::inferEncoding($snmp_data['ifAlias'] ?? null);
+
+            $port_id = get_port_id($ports_mapped, $snmp_data, $port_association_mode);
+
+            if (! $port_id) {
+                // If the SNMP to port_id lookup fails, we need to be aware of conflicts later
+                $ifindex_conflicts[$ifIndex] = $ports_ifindex[$ifIndex];
+            } elseif ($ports_ifindex[$ifIndex] != $port_id) {
+                // The SNMP data found an existing port that is different from the DB row with the same ifIndex
+                d_echo("port_id has changed for interface with index $ifIndex\n");
+
+                // Fetch the data of the keys we want to swap
+                $old_ifIndex = $ports_db[$port_id]['ifIndex'];
+                $conflicting_port_id = $ports_ifindex[$ifIndex];
+
+                // Use a transaction to swap the indexes.  We use null ifIndex temporarily to avoid duplicate key constraints
+                DB::transaction(function () use ($port_id, $ifIndex, $conflicting_port_id, $old_ifIndex) {
+                    DB::update('UPDATE ports SET ifIndex=NULL WHERE port_id=:port_id', ['port_id' => $conflicting_port_id]);
+                    DB::update('UPDATE ports SET ifIndex=:ifIndex WHERE port_id=:port_id', ['ifIndex' => $ifIndex, 'port_id' => $port_id]);
+                    DB::update('UPDATE ports SET ifIndex=:ifIndex WHERE port_id=:port_id', ['ifIndex' => $old_ifIndex, 'port_id' => $conflicting_port_id]);
+                });
+
+                // Update the maps for both ports
+                $ports_db[$conflicting_port_id]['ifIndex'] = $old_ifIndex;
+                $ports_ifindex[$old_ifIndex] = $conflicting_port_id;
+                $ports_db[$port_id]['ifIndex'] = $ifIndex;
+                $ports_ifindex[$ifIndex] = $port_id;
+            } elseif (array_key_exists($ifIndex, $ifindex_conflicts)) {
+                // We can just update this record because there is no conflict on the ifIndex, and another port wants to use the existing ifIndex
+                DB::update('UPDATE ports SET ifIndex=:ifIndex WHERE port_id=:port_id', ['ifIndex' => $ifIndex, 'port_id' => $port_id]);
+
+                // Update the maps
+                $ports_db[$port_id]['ifIndex'] = $ifIndex;
+                $ports_ifindex[$ifIndex] = $port_id;
+
+                // This is no longer a conflict
+                unset($ifindex_conflicts[$ifIndex]);
+            }
+        }
+    }
+
+    // Go through remaining conflicts and delete any rows marked as deleted
+    foreach ($ifindex_conflicts as $ifIndex => $port_id) {
+        DB::update('DELETE FROM ports WHERE deleted=1 AND port_id=:port_id', ['port_id' => $port_id]);
```

Comment:
Ports table only ever soft deletes ports.  We should not be deleting them here.

Reply:
I am assuming that the constraint exists for a reason, probably because the default configuration uses ifIndex as the port association mode.  Also, my code won't run unless someone changes the port association mode.

I did read an issue where someone was having problems with deleted ports stopping an active port from graphing, and I think that deleting a port is better than having an active port not graphing.

Alternate options I can think of are:
 - Take over the deleted port based on ifIndex instead of deleting it, but this could cause old unrelated data to appear on the interface graph
 - Remove the unique key, but I think this could cause issues with the default port association mode
 - Create a new unique key field which we update to match the value from the port association mode (and add something to make sure this field is synced to the correct value at the start of discovery in case the association mode changes)

If you think showing old data is better than deleting the row, I am happy to have a go at implementing something to do this instead.

Reply:
"I am assuming that the constraint exists for a reason" That is a big assumption :D

Reply:
I suspect the ifIndex constraint is a pre-fork relic.  Likely predates port association mode.

Reply:
> The goal should be to keep the port_id mapped to the same port. If a transient condition makes all ports be removed, we need to re-attach the ports when they are back.

I believe that this new code will still work correctly if all ports disappear and then re-appear, if ports all get new ifIndexes.  The only issue is if a small number of ports disappear, and one of them holds onto an ifIndex that has been re-assigned.  In this last scenario, it will still take 2 discovery cycles to delete a port (the first cycle will mark the port as deleted, then the second will delete the port).

I think that this is an acceptable compromise, especially since it's not modifying the behaviour of the system unless someone changes the port association mode.

Having said this, I have suggested 3 options above that I'm willing to implement if we really don't want to delete ports.